negotiation/neural/preprocess.py

```diff
@@ -413,6 +401,44 @@ def process_event(self, e, kb):
         else:
             raise ValueError('Unknown event action.')
 
+    def is_keyword(self, token):
+        summary_keywords = ["what", "who", "when", "where", "how", "price",
+            "can", "not", "cannot", "interested", "fair", "only", "would",
+            "low", "lower", "high", "higher", "good", "bad", "?", "!"]
+        return token in summary_keywords
+
+    def summarize(self, utterance):
+        '''
+        Keep only the special, summary keywords that signify user intent. We
+        purposely keep duplicates because having repeats can be a good signal,
+        also preservers order because "can not" and "not bad" are meaningful
+        '''
+        summary = []
+        for token in utterance:
+            if is_entity(token) or self.is_keyword(token):
+                summary.append(token)
+        return summary
+
+    def _process_example(self, ex):
+        '''
+        Convert example to turn-based dialogue from each agent's perspective
+        Create two Dialogue objects for each example
+        '''
+        kbs = ex.scenario.kbs
+        for agent in (0, 1):
+            dialogue = Dialogue(agent, kbs[agent], ex.ex_id)
+            for e in ex.events:
+                utterance = self.process_event(e, dialogue.kb)
+                if utterance:
+                    if self.model == "sum2sum":
+                        utterance = self.summarize(utterance)
+                    elif self.model == "sum2seq":
+                        summary = self.summarize(utterance)
+                        # Do we want to add some EOD marker here?
+                        summary.extend(utterance)
+                    dialogue.add_utterance(e.agent, utterance, lf=e.metadata)
```

Comment:
can you add summary separately, i.e. don't overwrite utterance, because for sum2seq you still need the utterance data.